internal/operator/reconciliations/api_gateway/reconciliation.go

```diff
@@ -0,0 +1,149 @@
+	Reconcile(ctx context.Context, apiGatewayCR operatorv1alpha1.APIGateway, apiGatewayResourceListPath string) (operatorv1alpha1.APIGateway, controllers.Status)
+}
+
+type Reconciliation struct {
+	Client client.Client
+}
+
+const (
+	reconciliationFinalizer string = "apigateways.operator.kyma-project.io/api-gateway-reconciliation"
+	dafaultGatewayName      string = "kyma-gateway"
+	dafaultGatewayNS        string = "kyma-system"
+)
+
+var checkDefaultGatewayReference = func(ctx context.Context, c client.Client, res resources.Resource) bool {
+	u := &unstructured.Unstructured{}
+	u.SetGroupVersionKind(res.GVK)
+
+	err := c.Get(ctx, client.ObjectKey{
+		Namespace: res.Namespace,
+		Name:      res.Name,
+	}, u)
+
+	if err != nil {
+		ctrl.Log.Error(err, "Error happened during obtaining user created object")
+	}
+
+	if res.GVK.Kind == "APIRule" && u.Object["spec"] != nil {
+		return u.Object["spec"].(map[string]interface{})["gateway"] == dafaultGatewayNS+"/"+dafaultGatewayName
+	} else if res.GVK.Kind == "VirtualService" && u.Object["spec"] != nil {
+		gateways := u.Object["spec"].(map[string]interface{})["gateways"]
+		if gateways != nil {
+			for _, gateway := range gateways.([]interface{}) {
+				if gateway == dafaultGatewayNS+"/"+dafaultGatewayName {
+					return true
+				}
+			}
+		}
+	}
+
+	return false
+}
+
+func (i *Reconciliation) Reconcile(ctx context.Context, apiGatewayCR operatorv1alpha1.APIGateway, apiGatewayResourceListPath string) (operatorv1alpha1.APIGateway, controllers.Status) {
+	if shouldDelete(apiGatewayCR) && hasReconciliationFinalizer(apiGatewayCR) {
+		ctrl.Log.Info("Starting API-Gateway deletion")
+
+		resourceFinder, err := resources.NewResourcesFinderFromConfigYaml(ctx, i.Client, ctrl.Log, apiGatewayResourceListPath)
+		if err != nil {
+			return apiGatewayCR, controllers.ErrorStatus(err, "Could not read customer resources finder configuration")
+		}
+
+		clientResources, err := resourceFinder.FindUserCreatedResources(checkDefaultGatewayReference)
+		if err != nil {
+			return apiGatewayCR, controllers.ErrorStatus(err, "Could not get customer resources from the cluster")
+		}
+
+		if len(clientResources) > 0 {
+			for _, res := range clientResources {
+				ctrl.Log.Info("Customer resource is blocking API-Gateway deletion", res.GVK.Kind, fmt.Sprintf("%s/%s", res.Namespace, res.Name))
+			}
+
+			return apiGatewayCR, controllers.WarningStatus(fmt.Errorf("could not delete API-Gateway module instance since there are %d custom resource(s) present that block its deletion", len(clientResources)),
+				"There are custom resource(s) that block the deletion. Please take a look at kyma-system/api-gateway-controller-manager logs to see more information about the warning")
+		}
+
+		//Temporarily having this simple solution for deletion of Kyma default gateway before we migrate whole management of Kyma default gateway from Istio component
+		if err := deleteDefaultGateway(ctx, i.Client); err != nil {
```

Comment:
I think we still need to move this to the gateway package. All logic about the Kyma gateway should be encapsulated there. We can expose a function to call it here. Also there is a Kyma Gateway finalizer that will not be handled correctly if deleted like this.